### PR TITLE
[onert] Optimize Bias Grad Computation

### PR DESCRIPTION
**Pull request by Aeren1564**
This commit optimizes bias grad computation with multithreading.

ONE-DCO-1.0-Signed-off-by: YongHyun An <yonghyunz.an@samsung.com>

---
From draft https://github.com/Samsung/ONE/pull/12571, https://github.com/Samsung/ONE/pull/12666
Relevant issue: https://github.com/Samsung/ONE/issues/12569

### CODE REVIEW

**Comment by Aeren1564** [comment]
Training Time for [mobilenet_v2](https://github.com/Samsung/ONE/issues/12325#issue-2047802908)


Before this commit
---
```asm
== training parameter ==
- learning_rate   = 0.001
- batch_size      = 32
- loss_info       = {loss = mean squared error, reduction = sum over batch size}
- optimizer       = adam
========================
Epoch 1/5 - time: 4430.210ms/step - loss: [0] 30090.0898
Epoch 2/5 - time: 4607.027ms/step - loss: [0] 30090.0840
Epoch 3/5 - time: 4669.991ms/step - loss: [0] 30090.0840
Epoch 4/5 - time: 4705.184ms/step - loss: [0] 30090.0840
Epoch 5/5 - time: 4659.798ms/step - loss: [0] 30090.0840
===================================
MODEL_LOAD   takes 5.4760 ms
PREPARE      takes 786.9770 ms
EXECUTE      takes 207755.0330 ms
- Epoch 1      takes 39871.8920 ms
- Epoch 2      takes 41463.2420 ms
- Epoch 3      takes 42029.9200 ms
- Epoch 4      takes 42346.6520 ms
- Epoch 5      takes 41938.1790 ms
===================================
```

After this commit
---
```asm
== training parameter ==
- learning_rate   = 0.001
- batch_size      = 32
- loss_info       = {loss = mean squared error, reduction = sum over batch size}
- optimizer       = adam
========================
Epoch 1/5 - time: 2800.052ms/step - loss: [0] 30090.0977
Epoch 2/5 - time: 2580.089ms/step - loss: [0] 30090.0938
Epoch 3/5 - time: 2654.771ms/step - loss: [0] 30090.0938
Epoch 4/5 - time: 2647.186ms/step - loss: [0] 30090.0938
Epoch 5/5 - time: 2648.930ms/step - loss: [0] 30090.0938
===================================
MODEL_LOAD   takes 5.3360 ms
PREPARE      takes 775.7780 ms
EXECUTE      takes 120092.1480 ms
- Epoch 1      takes 25200.4690 ms
- Epoch 2      takes 23220.8040 ms
- Epoch 3      takes 23892.9350 ms
- Epoch 4      takes 23824.6720 ms
- Epoch 5      takes 23840.3740 ms
===================================
```

File: runtime/onert/backend/train/ops/DepthwiseConvolutionLayer.cc [hunk 1]
```diff
@@ -170,17 +169,8 @@ void DepthwiseConvolutionLayer::backwardFloat32()
   // Calculate gradient for bias
   if (_bias)
   {
-    // TODO Use optimized kernel
```
Comment (Aeren1564) [inline runtime/onert/backend/train/ops/DepthwiseConvolutionLayer.cc:173]: This commit resolves this TODO item.

File: runtime/onert/backend/train/ops/ConvolutionLayer.cc [hunk 2]
```diff
@@ -196,20 +195,8 @@ void ConvolutionLayer::backwardFloat32()
   // Calculate gradient for bias
   if (_bias)
   {
-    // TODO Use optimized kernel
```
Comment (Aeren1564) [inline runtime/onert/backend/train/ops/ConvolutionLayer.cc:199]: This commit resolves this TODO item.

File: runtime/onert/backend/train/ops/OperationUtils.cc [hunk 3]
```diff
@@ -16,9 +16,184 @@
 
 #include "OperationUtils.h"
 
+#include <cker/eigen/EigenSupport.h>
+#include <cker/eigen/Utils.h>
+#include <cker/operation/Helper/Tensor.h>
 #include <cker/train/operation/ReLU.h>
 #include <cker/train/operation/ReLU6.h>
 
+// From tensorflow/core/kernels/redux_functor.h
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/OperationUtils.cc:25]: Please create a separate file for this change.

Comment (Aeren1564) [reply runtime/onert/backend/train/ops/OperationUtils.cc]: I've created a separate file named `ReduxFunctor.h`.

**Review by YongseopKim — COMMENTED** [review]
In `runtime/onert/backend`, there are directly using sths in `Eigen`. In convention, onert doesn't use directly specific kernel such as `Eigen` AFAIK. How about hiding this? 
- AS-IS: in `biasGrad()`, we can see `Eigen`.
- TO-BE: in `biasGradIO`, there are no `Eigen`.

I'd like to get feedback from others.

**Comment by jyoungyun** [comment]
@Aeren1564 Would you like to look at another file structure? This is because this PR doesn't fit with the existing code design. I think this  `reduxFunctor.h` should be located in the `compute/cker/include/cker/eigen/`. And when copying code from other repository, it is recommended not to modify it as much as possible. So it would be better to keep the file name as before. :)

File: runtime/onert/backend/train/ops/OperationUtils.cc [hunk 4]
```diff
@@ -64,6 +65,32 @@ const IPortableTensor *backpropActivation(const ir::Activation &activation,
...
+  nnfw::cker::Shape input_backprop_shape = getShape(input_backprop);
+  float *input_backprop_buffer = reinterpret_cast<float *>(input_backprop->buffer());
+  const nnfw::cker::Tensor input_backprop_t{input_backprop_shape,
+                                            static_cast<void *>(input_backprop_buffer)};
+
+  nnfw::cker::Shape bias_grad_shape = getShape(bias_grad);
+  float *bias_grad_buffer = getBuffer<float>(bias_grad);
+  nnfw::cker::Tensor bias_grad_t{bias_grad_shape, bias_grad_buffer};
+
+  Eigen::Index outer = input_backprop_shape.Dims(input_backprop_shape.DimensionsCount() - 1);
```
Comment (YongseopKim) [inline runtime/onert/backend/train/ops/OperationUtils.cc:85]: These `Eigen` class cannot be hidden?

Until now, onert has hidden these things AFAIK.

```
ONE/runtime/onert/backend (master ✗) grep -rni "eigen"
train/TensorManager.h:38:  // _align is equal to EIGEN_MAX_ALIGN_BYTES, though including Eigen headers
``

Comment (Aeren1564) [reply runtime/onert/backend/train/ops/OperationUtils.cc]: I hid "Eigen" from onert but now `redux_functor` accepts the following arguments

```cpp
void operator()(const Device &device, const Eigen::Index &outer_dim, const Eigen::Index &inner_dim, const Tensor &input, Tensor *output) const
```

Please let me know if you have some better way to handle this.

Comment (jyoungyun) [reply runtime/onert/backend/train/ops/OperationUtils.cc]: The built-in type can be transmitted and converted into the `Eigen::Index` in `redux_functor` function.

**Comment by YongseopKim** [comment]
I thought like [this](https://github.com/YongseopKim/ONE/commit/251b32c875095d161f90f7ab902c41cce2c7832b) @Aeren1564 .

File: runtime/onert/backend/train/ops/OperationUtils.cc (hunk 4 above)
Comment (Aeren1564) [reply runtime/onert/backend/train/ops/OperationUtils.cc]: > The built-in type can be transmitted and converted into the `Eigen::Index` in `redux_functor` function.

I implemented @YongseopKim 's suggestion for now :eyes:

File: compute/cker/include/cker/eigen/redux_functor.h [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+          buf0, buf);
+      }
+      // Write final result to the output.
+      output->template flat<OutputT>() = buf0.template cast<OutputT>().reshape(output_dims);
+    }
+  }
+};
+
+void biasReductionHelper(float *input_backprop_buffer, const Shape &input_backprop_shape,
+                         float *bias_grad_buffer, const Shape &bias_grad_shape)
```
Comment (Aeren1564) [inline compute/cker/include/cker/eigen/redux_functor.h:194]: This is from @YongseopKim 's [suggestion](https://github.com/Samsung/ONE/pull/12673#issuecomment-1963485619).

**Review by YongseopKim — APPROVED** [review]
LGTM 👍

**Review by jyoungyun — APPROVED** [review]
LGTM

**Pull request merged**